packages/cactus-api-client/src/main/typescript/socketio-api-client.ts

```diff
@@ -0,0 +1,345 @@
+  private monitorSubject: ReplaySubject<SocketLedgerEvent> | undefined;
+
+  readonly className: string;
+  counterReqID = 1;
+  checkValidator: (
+    key: string,
+    data: string,
+  ) => Promise<any> = verifyValidatorJwt;
+
+  constructor(public readonly options: SocketIOApiClientOptions) {
+    this.className = this.constructor.name;
+
+    Checks.nonBlankString(
+      options.validatorID,
+      `${this.className}::constructor() validatorID`,
+    );
+    Checks.nonBlankString(
+      options.validatorURL,
+      `${this.className}::constructor() validatorURL`,
+    );
+    Checks.nonBlankString(
+      // TODO - checks path exists?
+      options.validatorKeyPath,
+      `${this.className}::constructor() validatorKeyPath`,
+    );
+
+    const level = this.options.logLevel || "INFO";
+    const label = this.className;
+    this.log = LoggerProvider.getOrCreate({ level, label });
+
+    this.log.info(
+      `Created ApiClient for Validator ID: ${options.validatorID}, URL ${options.validatorURL}, KeyPath ${options.validatorKeyPath}`,
+    );
+    this.log.debug("socketOptions:", options.socketOptions);
+
+    this.socket = io(options.validatorURL, options.socketOptions);
+  }
+
+  public sendAsyncRequest(
+    contract: Record<string, unknown>,
+    method: Record<string, unknown>,
+    args: any,
+  ): void {
+    try {
+      const requestData = {
+        contract: contract,
+        method: method,
+        args: args,
+      };
+
+      this.log.debug("sendAsyncRequest() Request:", requestData);
+      this.socket.emit("request2", requestData);
+    } catch (err) {
+      this.log.error("sendAsyncRequest() EXCEPTION", err);
+      throw err;
+    }
+  }
+
+  // TODO - Refactor to RxJS
+  public sendSyncRequest(
+    contract: Record<string, unknown>,
+    method: Record<string, unknown>,
+    args: any,
+  ): Promise<any> {
+    return new Promise((resolve, reject) => {
+      this.log.debug("call : sendSyncRequest");
+
+      try {
+        this.log.debug(
+          "##in sendSyncRequest, contract:",
+          contract,
+          "method:",
+          method,
+          "args:",
+          args,
+        );
+        let responseFlag = false;
+
+        // reqID generation
+        const reqID = this.genarateReqID();
+        this.log.debug(`##sendSyncRequest, reqID = ${reqID}`);
+
+        this.socket.on("connect_error", (err: Error) => {
+          this.log.error("##connect_error:", err);
+          this.socket.disconnect();
+          reject(err);
+        });
+        this.socket.on("connect_timeout", (err: Record<string, unknown>) => {
+          this.log.error("####Error:", err);
+          this.socket.disconnect();
+          reject(err);
+        });
+        this.socket.on("error", (err: Record<string, unknown>) => {
+          this.log.error("####Error:", err);
+          this.socket.disconnect();
+          reject(err);
+        });
+        this.socket.on("response", (result: any) => {
```

Comment:
@outSH How do different types of responses get disambiguted here? Is this assumption baked in that the entire system only supports one type of response for a single use-case pertaining to a single endpoint/backend logic?

Reply:
@petermetz socketio-server At the moment Verifier is meant to be used in a backend only, I've adjusted webpack imports to reflect that (so browser package will not contain these functions). Also, I've mentioned in the github issue that I plan adding browser support in the future, but definitely in separate PR.

Reply:
@outSH What I mean is how would something like this be possible on the same socket connection?
Would you do routing based on the payload of the message with some lookup at runtime for the correct handler?

```typescript

const someLogic = () => {
// do something
}

const someOtherDifferentLogic = () => {
// do something else
}

socket.on("request", someLogic);
socket.on("request", someOtherDifferentLogic);
```